stdlib/SuiteSparse/src/umfpack.jl

```diff
@@ -428,57 +428,208 @@ for itype in UmfpackIndexTypes
+                @isok ccall(($get_num_r, :libumfpack), $itype,
+                            (Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cvoid},
+                             Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cvoid},
+                             Ptr{$itype}, Ptr{Cvoid}, Ptr{Cvoid},
+                             Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cvoid}),
+                            C_NULL, C_NULL, C_NULL,
+                            C_NULL, C_NULL, C_NULL,
+                            P, C_NULL, C_NULL,
+                            C_NULL, C_NULL, lu.numeric)
+                return increment!(P)
+            elseif d === :q
+                Q  = Vector{$itype}(undef, n_col)
+                @isok ccall(($get_num_r, :libumfpack), $itype,
+                            (Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cvoid},
+                             Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cvoid},
+                             Ptr{Cvoid}, Ptr{$itype}, Ptr{Cvoid},
+                             Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cvoid}),
+                            C_NULL, C_NULL, C_NULL,
+                            C_NULL, C_NULL, C_NULL,
+                            C_NULL, Q, C_NULL,
+                            C_NULL, C_NULL, lu.numeric)
+                return increment!(Q)
+            elseif d === :Rs
+                Rs = Vector{Float64}(undef, n_row)
+                @isok ccall(($get_num_r, :libumfpack), $itype,
+                            (Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cvoid},
+                             Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cvoid},
+                             Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cvoid},
+                             Ptr{Cvoid}, Ptr{Float64}, Ptr{Cvoid}),
+                            C_NULL, C_NULL, C_NULL,
+                            C_NULL, C_NULL, C_NULL,
+                            C_NULL, C_NULL, C_NULL,
+                            C_NULL, Rs, lu.numeric)
+                return Rs
+            elseif d === :(:)
+                Lp = Vector{$itype}(undef, n_row + 1)
+                # L is returned in CSR (compressed sparse row) format
+                Lj = Vector{$itype}(undef, lnz)
+                Lx = Vector{Float64}(undef, lnz)
+                Up = Vector{$itype}(undef, n_col + 1)
+                Ui = Vector{$itype}(undef, unz)
+                Ux = Vector{Float64}(undef, unz)
+                P  = Vector{$itype}(undef, n_row)
+                Q  = Vector{$itype}(undef, n_col)
+                Rs = Vector{Float64}(undef, n_row)
+                @isok ccall(($get_num_r, :libumfpack), $itype,
+                            (Ptr{$itype}, Ptr{$itype}, Ptr{Float64},
+                             Ptr{$itype}, Ptr{$itype}, Ptr{Float64},
+                             Ptr{$itype}, Ptr{$itype}, Ptr{Cvoid},
+                             Ptr{Cvoid}, Ptr{Float64}, Ptr{Cvoid}),
+                            Lp, Lj, Lx,
+                            Up, Ui, Ux,
+                            P, Q, C_NULL,
+                            C_NULL, Rs, lu.numeric)
```

Comment:
Not that I understand much/anything here, so this may be totally stupid, but I did notice that compared to the original code above the `C_NULL` in the last quoted line used to be a `0`, and the corresponding type in the signature a few lines earlier was `Ref{$itype}`. Is that intentional? I guess it is, because in all branches this is a `C_NULL` now.

Reply:
According to the documentation, this is an output argument (like Lp, Lj...) so it is optional and passing C_NULL avoid it being copied. Since we don't use it, it seems that C_NULL is the best and only sensible choice for this parameter. I can't figure out why it was set to `0` while `Dx` (the preceding parameter) is correctly set to C_NULL. My best guess is that it was mistaken for an input parameter when the wrapper was first written.